src/store_client.cc

```diff
@@ -499,58 +517,126 @@ store_client::fileRead()
-    if (copyInto.offset == 0 && len > 0 && rep && rep->sline.status() == Http::scNone) {
-        /* Our structure ! */
-        if (!entry->mem_obj->adjustableBaseReply().parseCharBuf(copyInto.data, headersEnd(copyInto.data, len))) {
-            debugs(90, DBG_CRITICAL, "ERROR: Could not parse headers from on disk object");
-        }
+    if (!lastIoResult) {
+        if (answeredOnce())
+            return noteEof();
+
+        debugs(90, DBG_CRITICAL, "ERROR: Truncated HTTP headers in on-disk object");
+        return fail();
     }
 
-    if (len > 0 && rep && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
-        storeGetMemSpace(len);
+    assert(lastRead_.data == buf);
+    lastRead_.length = lastIoResult;
+
+    parsingBuffer->appended(buf, lastIoResult);
+
+    maybeWriteFromDiskToMemory(lastRead_);
+    handleBodyFromDisk();
+}
+
+/// de-serializes HTTP response (partially) read from disk storage
+void
+store_client::handleBodyFromDisk()
+{
+    // We cannot de-serialize on-disk HTTP response without MemObject because
+    // without MemObject::swap_hdr_sz we cannot know where that response starts.
+    Assure(entry->mem_obj);
+    Assure(entry->mem_obj->swap_hdr_sz > 0);
+
+    if (!answeredOnce()) {
+        // All on-disk responses have HTTP headers. First disk body read(s)
+        // include HTTP headers that we must parse (if needed) and skip.
+        const auto haveHttpHeaders = entry->mem_obj->baseReply().pstate == Http::Message::psParsed;
+        if (!haveHttpHeaders && !parseHttpHeadersFromDisk())
+            return;
+        skipHttpHeadersFromDisk();
+    }
+
+    // no callback(): we handled negative/zero lastIoResult ourselves
+    noteNews();
+}
+
+/// Adds HTTP response data loaded from disk to the memory cache (if
+/// needed/possible). The given part may contain portions of HTTP response
+/// headers and/or HTTP response body.
+void
+store_client::maybeWriteFromDiskToMemory(const StoreIOBuffer &httpResponsePart)
+{
+    // XXX: Reject [memory-]uncachable/unshareable responses instead of assuming
+    // that any HTTP response loaded from disk should be written to MemObject's
+    // data_hdr (and that it may purge already cached entries). Cachability
+    // decision(s) should be made outside (and obeyed by) this low-level code.
+    if (httpResponsePart.length && entry->mem_obj->inmem_lo == 0 && entry->objectLen() <= (int64_t)Config.Store.maxInMemObjSize && Config.onoff.memory_cache_disk) {
+        storeGetMemSpace(httpResponsePart.length);
+        // XXX: The "recheck" below is not needed because nobody can purge
+        // mem_hdr bytes of a locked entry, and we do lock our entry. Moreover,
+        // inmem_lo offset itself should not be relevant to appending new bytes.
         // recheck for the above call may purge entry's data from the memory cache
         if (entry->mem_obj->inmem_lo == 0) {
```

Comment:
Do you agree with this XXX? This question has a medium priority.

Reply:
I will replace "nobody" with storeGetMemSpace() to fix this TODO comment. You are right that purging certain bytes of locked entries is allowed -- that is how Squid's "read ahead gap" works -- but, in this specific context, we only care about storeGetMemSpace() actions, and those actions do not affect locked entries.

Reply:
Done in 6a3f522.